### ISSUE TITLE
Is it possible to implement a fast SFU for screen sharing?

### ISSUE THREAD

**Issue by NewMySQL** [sipsorcery-org/sipsorcery#1202]
I tried to create a simple SFU for or 2 clients using sipsorcery using:
1. Example [WebRTCExamples/WebRTCAspNetMvc](https://github.com/sipsorcery-org/sipsorcery/tree/master/examples/WebRTCExamples/WebRTCAspNetMvc)
2. Way from [Issue #1180](https://github.com/sipsorcery-org/sipsorcery/issues/1180)
Add event handler and change track in `WebRTCHostedService.cs`
```csharp
MediaStreamTrack videoTack = new MediaStreamTrack([
    new VideoFormat(VideoCodecsEnum.VP8, 96),
    new VideoFormat(VideoCodecsEnum.H264, 100)
], MediaStreamStatusEnum.SendRecv);
peerConnection.addTrack(videoTack);

peerConnection.OnVideoFrameReceived += (ep, timestamp, sample, fmt) =>
{
    foreach (var (pId, pC) in _peerConnections)
    {
        if (pId == id) continue;

        pC.SendVideo(timestamp, sample);
    }
};
```
Modify `Views/Home/Index.cshtml`
```cshtml
<script type="text/javascript">
    const id = '@Guid.NewGuid().ToString()';
    const baseUrl = "api/webrtc/";
    const getOfferUrl = `${baseUrl}getoffer?id=${id}`;
    const setAnswerUrl = `${baseUrl}setanswer?id=${id}`;
    const setIceCandidateUrl = `${baseUrl}addicecandidate?id=${id}`

    var pc

    async function start() {
        closePeer();

        let videoControl = document.querySelector('#videoSrc');
        let videoRecvControl = document.querySelector('#videoRecv');
        
        const localStream = await navigator.mediaDevices.getDisplayMedia({
            video: {
                frameRate: 60,
                width: 1920,
                height: 1080,
            }
         });
        videoControl.srcObject = localStream;

        pc = new RTCPeerConnection(null);
        
        localStream.getTracks().forEach(track => {
            console.log('add local track ' + track.kind + ' to peer connection.');
            console.log(track);
            pc.addTrack(track, localStream);
        });

        pc.ontrack = evt => {
            videoRecvControl.srcObject = evt.streams[0];
        }

        pc.onicegatheringstatechange = function () {
            console.log("onicegatheringstatechange: " + pc.iceGatheringState);
        }

        pc.oniceconnectionstatechange = function () {
            console.log("oniceconnectionstatechange: " + pc.iceConnectionState);
        }

        pc.onsignalingstatechange = function () {
            console.log("onsignalingstatechange: " + pc.signalingState);
        }

        pc.onicecandidate = async function (event) {
            if (event.candidate) {
                console.log('new-ice-candidate:');
                console.log(event.candidate.candidate);
                console.log(event.candidate);
                await fetch(setIceCandidateUrl, {
                    method: 'POST',
                    body: JSON.stringify(event.candidate),
                    headers: { 'Content-Type': 'application/json' }
                });
            }
        };

        let offerResponse = await fetch(getOfferUrl);
        let offer = await offerResponse.json();
        console.log("got offer: " + offer.type + " " + offer.sdp + ".");
        await pc.setRemoteDescription(offer);

        pc.createAnswer().then(function (answer) {
            return pc.setLocalDescription(answer);
        }).then(async function () {
            console.log("Sending answer SDP.");
            console.log("SDP: " + pc.localDescription.sdp);
            await fetch(setAnswerUrl, {
                method: 'POST',
                body: JSON.stringify(pc.localDescription),
                headers: { 'Content-Type': 'application/json' }
            });
        });

    }

    function closePeer() {
        if (pc != null) {
            console.log("close peer");
            pc.close();
        }
    }
</script>

<video controls autoplay="autoplay" id="videoSrc" width="480"></video>
<video controls autoplay="autoplay" id="videoRecv" width="480"></video>

<div>
    <button type="button" class="btn btn-success" onclick="start();">Start</button>
    <button type="button" class="btn btn-success" onclick="closePeer();">Close</button>
</div>
```

But got problems:
1) The second connected client loads the video stream in about ~60 seconds, although the first client does not have such problems, instant loading
2) On another computer, even in a local network, the `oniceconnectionstatechange: connected` takes about 5-10 seconds.
3) VP8: The quality of the transmitted video is very terrible, and most of the frames are lost, very low fps.
    H264: Frequent freezing and subsequent accelerations. 

P.s. there are no problems with the speed and stability of the local connection, a similar simple SFU using mediasoup does not cause such problems

**Comment by ha-ves** [comment]
I replicated this issue via static file serving and off-band signaling but only had some of the same problems.

1. It pretty much connects instantly after signaling,
1. H264 via FFmpeg, not so much "freezing" but it does "delay and accelerate" (as in keyframe jumps)

I'm guessing this might be on the RTP part and probably a timestamp issue causing the H264 keyframe jumps.

**Comment by sipsorcery** [comment]
It's worth pointing out that the original purpose from this libray, as you can probably guess, was a SIP stack. SIP is a signalling only protocol, i.e. no RTP, mo media etc. Eventually a single class rudimentary RTP implementation was added to help with testing some SIP scenarios.

Then a decade and a bit further down the track when WebRTC started to get popular and I, and some others, had some spare time we had a crack at getting WebRTC working. Again in order to do some useful testing some rudimentary audio & video encoding & decoding was useful. At the time it was a major achievement just to get a WebRTC connection with a browser. Getting some video & audio flowing was the icing on top.

tldr; For this library audio, video and performance in general has never been a focus. To use this library, or even any managed runtime language, for a high performance use case like an SFU you'll need to spend a lot of time milking performance. Personally I'd just go straight to C++.
